mqtt/mqtt-broker/src/server.rs

```diff
@@ -56,36 +68,85 @@ where
+
+                        for e in results.into_iter().filter_map(Result::err) {
+                            warn!(message = "failed to shutdown protocol head", error=%e);
+                        }
+
                         debug!("sending Shutdown message to broker");
                         handle.send(Message::System(SystemEvent::Shutdown)).await?;
                         broker_task.await.context(ErrorKind::TaskJoin)?
                     }
-                    Either::Left((broker_state, incoming_task)) => {
+                    Either::Left((broker_state, incoming_tasks)) => {
                         warn!("broker exited before accept loop");
-                        incoming_task.await?;
+
+                        // wait until either of incoming_tasks finished
+                        let (result, _index, unfinished_incoming_tasks) = incoming_tasks.await;
+
+                        // wait until the rest of incoming_tasks finished
+                        let mut results = vec![result];
+                        results.extend(future::join_all(unfinished_incoming_tasks).await);
+
+                        for e in results.into_iter().filter_map(Result::err) {
+                            warn!(message = "failed to shutdown protocol head", error=%e);
+                        }
+
                         broker_state.context(ErrorKind::TaskJoin)?
                     }
                 }
             }
-            Either::Right((either, _shutdown)) => match either {
-                Either::Right((Ok(_incoming_task), broker_task)) => {
-                    debug!("sending Shutdown message to broker");
-                    handle.send(Message::System(SystemEvent::Shutdown)).await?;
-                    broker_task.await.context(ErrorKind::TaskJoin)?
-                }
-                Either::Right((Err(e), broker_task)) => {
-                    error!(message = "an error occurred in the accept loop", error=%e);
+            Either::Right((either, _)) => match either {
+                Either::Right(((result, index, unfinished_incoming_tasks), broker_task)) => {
                     debug!("sending Shutdown message to broker");
+
+                    if let Err(e) = &result {
+                        error!(message = "an error occurred in the accept loop", error=%e);
+                    }
+
+                    debug!("sending stop signal for the rest of protocol heads");
+                    shutdown_handles.remove(index);
+                    send_shutdown(shutdown_handles);
+
+                    let mut results = vec![result];
+                    results.extend(future::join_all(unfinished_incoming_tasks).await);
+
                     handle.send(Message::System(SystemEvent::Shutdown)).await?;
-                    broker_task.await.context(ErrorKind::TaskJoin)?;
-                    return Err(e);
+
+                    let broker_state = broker_task.await;
+
+                    // todo refactor to possible gather all errors
```

Comment:
What about this TODO?

Reply:
now it logs the errors and passes back the broker state as in case of the rest of the error cases